File: agents/scheduler.py

from __future__ import annotations
from datetime import datetime, timedelta, date, time
from typing import List, Tuple, Dict, Optional
from core.models import Task, DayPlan, Block

Slot = Tuple[datetime, datetime]


def mock_energy_curve(day: date) -> List[Tuple[datetime, float]]:
    """
    Returns 96 (15-min) points across the day with simple peaks:
      - High:        09:00–11:00
      - Post-lunch:  13:00–14:00 dip
      - Medium-high: 16:00–18:00
    Values in [0..1].
    """
    base = datetime.combine(day, time(6, 0))
    curve: List[Tuple[datetime, float]] = []
    for i in range(96):
        t = base + timedelta(minutes=15 * i)
        hour = t.hour + t.minute / 60.0
        e = 0.3
        if 9 <= hour <= 11:
            e = 0.9
        if 16 <= hour <= 18:
            e = max(e, 0.8)
        if 13 <= hour <= 14:
            e = 0.2
        curve.append((t, e))
    return curve


def mock_busy(day: date) -> List[Slot]:
    """
    Hardcoded conflicts to simulate meetings:
      - 10:30–11:00
      - 14:00–15:00
    """
    s1 = datetime.combine(day, time(10, 30)); e1 = s1 + timedelta(minutes=30)
    s2 = datetime.combine(day, time(14, 0));  e2 = s2 + timedelta(minutes=60)
    return [(s1, e1), (s2, e2)]


# ----------------------- Slot / Grid Utilities -----------------------

def _workday_slots(day: date, start_h=9, end_h=18, step_min=15) -> List[Slot]:
    start = datetime.combine(day, time(start_h, 0))
    end = datetime.combine(day, time(end_h, 0))
    slots: List[Slot] = []
    t = start
    while t < end:
        nxt = t + timedelta(minutes=step_min)
        slots.append((t, nxt))
        t = nxt
    return slots


def _overlaps(a: Slot, b: Slot) -> bool:
    return not (a[1] <= b[0] or b[1] <= a[0])


def _free_slots(day: date, busy: List[Slot], step_min=15) -> List[Slot]:
    all_slots = _workday_slots(day, step_min=step_min)
    return [s for s in all_slots if not any(_overlaps(s, b) for b in busy)]


def _slot_scores(free: List[Slot], curve: List[Tuple[datetime, float]]) -> Dict[Slot, float]:
    energy_lookup = {t: e for t, e in curve}
    return {s: energy_lookup.get(s[0], 0.5) for s in free}


def _chunk_minutes_for_effort(effort: Optional[str]) -> int:
    """
    Choose chunk size by effort:
      - high: 60m
      - medium: 45m
      - low: 30m
    """
    if (effort or "").lower() == "high":
        return 60
    if (effort or "").lower() == "low":
        return 30
    return 45


def _contiguous_slots(start: datetime, minutes: int) -> List[Slot]:
    """Return the list of 15-min slots covering [start, start+minutes]."""
    steps = max(1, minutes // 15)
    return [
        (start + timedelta(minutes=15 * i),
         start + timedelta(minutes=15 * (i + 1)))
        for i in range(steps)
    ]


def _snap_down_15(dt: datetime) -> datetime:
    minutes = (dt.minute // 15) * 15
    return dt.replace(minute=minutes, second=0, microsecond=0)


def _snap_up_15(dt: datetime) -> datetime:
    if dt.minute % 15 == 0 and dt.second == 0 and dt.microsecond == 0:
        return dt.replace(second=0, microsecond=0)
    delta = 15 - (dt.minute % 15)
    out = dt + timedelta(minutes=delta)
    return out.replace(second=0, microsecond=0)


def _merge_adjacent(plan: DayPlan) -> DayPlan:
    """Fuse adjacent blocks of the same task to reduce fragmentation."""
    merged: List[Block] = []
    for b in sorted(plan.blocks, key=lambda x: (x.start, x.task_title)):
        if merged and merged[-1].task_title == b.task_title and merged[-1].end == b.start:
            merged[-1].end = b.end
        else:
            merged.append(b)
    plan.blocks = merged
    return plan




def greedy_schedule(
    tasks: List[Task],
    day: date,
    *,
    energy_curve: Optional[List[Tuple[datetime, float]]] = None,
    busy: Optional[List[Slot]] = None,
    work_start_h: int = 9,
    work_end_h: int = 18,
    step_min: int = 15,
) -> DayPlan:
    """
    Greedy packer with support for fixed-time tasks.

    Steps:
      1) Reserve fixed-time tasks first (fixed_start/fixed_end).
      2) Score remaining free slots against the energy curve.
         - high effort → chase peaks
         - low effort  → prefer dips (inverse score)
         - medium      → mild preference for energy
      3) Place tasks in contiguous chunks (30–60m) until done or grid full.
      4) Merge adjacent blocks of the same task.
    """
    curve = energy_curve or mock_energy_curve(day)
    busy_list = busy or mock_busy(day)

   
    all_work_slots = _workday_slots(day, start_h=work_start_h, end_h=work_end_h, step_min=step_min)
    free = [s for s in all_work_slots if not any(_overlaps(s, b) for b in busy_list)]
    scores = _slot_scores(free, curve)

    plan = DayPlan(date=day, blocks=[])
    used: set[Slot] = set()

    
    fixed, flexible = [], []
    for t in tasks:
        fs = getattr(t, "fixed_start", None)
        fe = getattr(t, "fixed_end", None)
        if fs or fe:
            
            if fs and not fe:
                fe = fs + timedelta(minutes=int(max(15, t.est_minutes)))
            
            fs = _snap_down_15(fs) if fs else None
            fe = _snap_up_15(fe) if fe else None
            
            try:
                t.fixed_start, t.fixed_end = fs, fe  
            except Exception:
                pass
            fixed.append((t, fs, fe))
        else:
            flexible.append(t)

    for t, fs, fe in fixed:
        if not fs or not fe:
            continue
    
        if fs.date() != day:
            continue
        minutes = int((fe - fs).total_seconds() // 60)
        needed = _contiguous_slots(fs, minutes)
        for ns in needed:
            used.add(ns)
        plan.blocks.append(Block(task_title=t.title, start=fs, end=fe))


  
    free = [s for s in free if s not in used]
    if not free and len(plan.blocks) == 0:
        return plan  
    def task_key(t: Task):
        eff_rank = {"high": 0, "medium": 1, "low": 2}.get((t.effort or "medium").lower(), 1)
        d = t.deadline or datetime.combine(day, time(23, 59))
        return (eff_rank, d)

    tasks_sorted = sorted(flexible, key=task_key)

    def score_for_task(slot: Slot, effort: Optional[str]) -> float:
        e = scores.get(slot, 0.5)
        eff = (effort or "medium").lower()
        if eff == "high":
            return e                 
        if eff == "low":
            return 1.0 - e           
        return 0.5 + 0.5 * e         

    
    free_sorted_cache = None

    for t in tasks_sorted:
        remaining = max(15, int(t.est_minutes))
        chunk = _chunk_minutes_for_effort(t.effort)

   
        free_sorted = sorted(free, key=lambda x: -score_for_task(x, t.effort))
        if free_sorted_cache is None:
            free_sorted_cache = free_sorted  

        while remaining > 0:
            placed_any = False
            for s in free_sorted:
                if s in used:
                    continue

                
                minutes = min(chunk, remaining)
                if minutes < 30 and remaining >= 30:
                    minutes = 30

             
                start = s[0]
                needed = _contiguous_slots(start, minutes)
                if all(ns in free and ns not in used for ns in needed):
                    for ns in needed:
                        used.add(ns)
                    plan.blocks.append(Block(task_title=t.title, start=start, end=start + timedelta(minutes=minutes)))
                    remaining -= minutes
                    placed_any = True
                    break  

            if not placed_any:
                
                break

    plan.blocks.sort(key=lambda b: b.start)
    _merge_adjacent(plan)
    return plan




def energy_alignment(plan: DayPlan) -> float:
    """
    Percent of scheduled minutes that fall in high-energy windows (>= 0.75).
    Returns a value in [0,1], rounded to 2 decimals.
    """
    if not plan.blocks:
        return 0.0
    curve = dict(mock_energy_curve(plan.date))
    hi_thresh = 0.75
    hi = 0
    total = 0
    for b in plan.blocks:
        t = b.start.replace(second=0, microsecond=0)
        while t < b.end:
            total += 15
            if curve.get(t, 0.0) >= hi_thresh:
                hi += 15
            t += timedelta(minutes=15)
    return 0.0 if total == 0 else round(hi / total, 2)

File: bootcamp-boiler/lib/agents/scheduler.ts
// lib/agents/scheduler.ts - Energy-aware task scheduling

import { Task, DayPlan, Block, Slot, SlotPoint } from '../types';

function mockBusy(day: Date): Slot[] {
  // Simulate some blocked times
  const start1 = new Date(day);
  start1.setHours(10, 30, 0, 0);
  const end1 = new Date(start1.getTime() + 30 * 60 * 1000);
  
  const start2 = new Date(day);
  start2.setHours(14, 0, 0, 0);
  const end2 = new Date(start2.getTime() + 60 * 60 * 1000);
  
  return [[start1, end1], [start2, end2]];
}

function workdaySlots(day: Date, startH: number = 9, endH: number = 18, stepMin: number = 15): Slot[] {
  const start = new Date(day);
  start.setHours(startH, 0, 0, 0);
  
  const end = new Date(day);
  end.setHours(endH, 0, 0, 0);
  
  const slots: Slot[] = [];
  let t = new Date(start);
  
  while (t < end) {
    const next = new Date(t.getTime() + stepMin * 60 * 1000);
    slots.push([new Date(t), new Date(next)]);
    t = next;
  }
  
  return slots;
}

function overlaps(a: Slot, b: Slot): boolean {
  return !(a[1] <= b[0] || b[1] <= a[0]);
}

function freeSlots(day: Date, busy: Slot[], stepMin: number = 15): Slot[] {
  const allSlots = workdaySlots(day, 9, 18, stepMin);
  return allSlots.filter(s => !busy.some(b => overlaps(s, b)));
}

function slotScores(free: Slot[], curve: SlotPoint[]): Map<Slot, number> {
  const energyLookup = new Map(curve.map(([t, e]) => [t.getTime(), e]));
  const scores = new Map<Slot, number>();
  
  for (const slot of free) {
    const energy = energyLookup.get(slot[0].getTime()) || 0.5;
    scores.set(slot, energy);
  }
  
  return scores;
}

function chunkMinutesForEffort(effort?: string): number {
  switch (effort?.toLowerCase()) {
    case 'high': return 60;
    case 'low': return 30;
    default: return 45;
  }
}

function contiguousSlots(start: Date, minutes: number): Slot[] {
  const steps = Math.max(1, Math.floor(minutes / 15));
  const slots: Slot[] = [];
  
  for (let i = 0; i < steps; i++) {
    const slotStart = new Date(start.getTime() + i * 15 * 60 * 1000);
    const slotEnd = new Date(start.getTime() + (i + 1) * 15 * 60 * 1000);
    slots.push([slotStart, slotEnd]);
  }
  
  return slots;
}

function snapDown15(dt: Date): Date {
  const result = new Date(dt);
  const minutes = Math.floor(result.getMinutes() / 15) * 15;
  result.setMinutes(minutes, 0, 0);
  return result;
}

function snapUp15(dt: Date): Date {
  const result = new Date(dt);
  if (result.getMinutes() % 15 === 0 && result.getSeconds() === 0 && result.getMilliseconds() === 0) {
    return result;
  }
  const delta = 15 - (result.getMinutes() % 15);
  result.setMinutes(result.getMinutes() + delta, 0, 0);
  return result;
}

function clampToWorkday(dt: Date, day: Date, startH: number, endH: number): Date {
  const start = new Date(day);
  start.setHours(startH, 0, 0, 0);
  
  const end = new Date(day);
  end.setHours(endH, 0, 0, 0);
  
  return new Date(Math.max(Math.min(dt.getTime(), end.getTime()), start.getTime()));
}

function mergeAdjacent(plan: DayPlan): DayPlan {
  const merged: Block[] = [];
  const sorted = [...plan.blocks].sort((a, b) => a.start.getTime() - b.start.getTime());
  
  for (const block of sorted) {
    const last = merged[merged.length - 1];
    if (last && last.task_title === block.task_title && last.end.getTime() === block.start.getTime()) {
      last.end = block.end;
    } else {
      merged.push(block);
    }
  }
  
  return { ...plan, blocks: merged };
}

export function greedySchedule(
  tasks: Task[],
  day: Date,
  energyCurve?: SlotPoint[],
  busy?: Slot[],
  workStartH: number = 9,
  workEndH: number = 18,
  stepMin: number = 15
): DayPlan {
  const curve = energyCurve || [];
  const busyList = busy || mockBusy(day);
  
  // Build available slots
  const allWorkSlots = workdaySlots(day, workStartH, workEndH, stepMin);
  const free = allWorkSlots.filter(s => !busyList.some(b => overlaps(s, b)));
  const scores = slotScores(free, curve);
  
  const plan: DayPlan = { date: day, blocks: [] };
  const used = new Set<string>();
  
  // Handle fixed-time tasks first
  const fixed: Array<[Task, Date?, Date?]> = [];
  const flexible: Task[] = [];
  
  for (const task of tasks) {
    if (task.fixed_start || task.fixed_end) {
      let fs = task.fixed_start ? snapDown15(task.fixed_start) : undefined;
      let fe = task.fixed_end ? snapUp15(task.fixed_end) : undefined;
      
      if (fs && !fe) {
        fe = new Date(fs.getTime() + Math.max(15, task.est_minutes) * 60 * 1000);
      }
      
      fixed.push([task, fs, fe]);
    } else {
      flexible.push(task);
    }
  }
  
  // Schedule fixed tasks
  for (const [task, fs, fe] of fixed) {
    if (!fs || !fe) continue;
    if (fs.toDateString() !== day.toDateString()) continue;
    
    const clampedStart = clampToWorkday(fs, day, workStartH, workEndH);
    const clampedEnd = clampToWorkday(fe, day, workStartH, workEndH);
    
    if (clampedEnd <= clampedStart) continue;
    
    const minutes = Math.floor((clampedEnd.getTime() - clampedStart.getTime()) / (60 * 1000));
    const needed = contiguousSlots(clampedStart, minutes);
    
    for (const [start, end] of needed) {
      used.add(`${start.getTime()}-${end.getTime()}`);
    }
    
    plan.blocks.push({
      task_title: task.title,
      start: clampedStart,
      end: clampedEnd
    });
  }
  
  // Filter out used slots
  const availableSlots = free.filter(s => !used.has(`${s[0].getTime()}-${s[1].getTime()}`));
  
  if (availableSlots.length === 0 && plan.blocks.length === 0) {
    return plan;
  }
  
  // Sort flexible tasks by deadline, then effort
  const tasksSorted = [...flexible].sort((a, b) => {
    const aDeadline = a.deadline || new Date(day.getTime() + 24 * 60 * 60 * 1000);
    const bDeadline = b.deadline || new Date(day.getTime() + 24 * 60 * 60 * 1000);
    
    if (aDeadline.getTime() !== bDeadline.getTime()) {
      return aDeadline.getTime() - bDeadline.getTime();
    }
    
    const effortRank = { high: 0, medium: 1, low: 2 };
    const aRank = effortRank[a.effort || 'medium'];
    const bRank = effortRank[b.effort || 'medium'];
    return aRank - bRank;
  });
  
  // Score function for task-effort matching
  function scoreForTask(slot: Slot, effort?: string): number {
    const e = scores.get(slot) || 0.5;
    const h = slot[0].getHours();
    
    switch (effort?.toLowerCase()) {
      case 'high':
        return e; // High effort wants high energy
      case 'low':
        const earlyPenalty = h < Math.max(workStartH + 1, 9) ? 0.2 : 0.0;
        return (1.0 - e) - earlyPenalty; // Low effort prefers low energy
      default:
        return 0.5 + 0.5 * e; // Medium effort is flexible
    }
  }
  
  // Schedule flexible tasks
  for (const task of tasksSorted) {
    let remaining = Math.max(15, task.est_minutes);
    const chunk = chunkMinutesForEffort(task.effort);
    
    let latestEnd: Date | undefined;
    if (task.deadline && task.deadline.toDateString() === day.toDateString()) {
      latestEnd = task.deadline;
    }
    
    const freeSorted = [...availableSlots]
      .filter(s => !used.has(`${s[0].getTime()}-${s[1].getTime()}`))
      .sort((a, b) => scoreForTask(b, task.effort) - scoreForTask(a, task.effort));
    
    while (remaining > 0) {
      let placedAny = false;
      
      for (const slot of freeSorted) {
        if (used.has(`${slot[0].getTime()}-${slot[1].getTime()}`)) continue;
        
        let minutes = Math.min(chunk, remaining);
        if (minutes < 30 && remaining >= 30) {
          minutes = 30;
        }
        
        const start = slot[0];
        const end = new Date(start.getTime() + minutes * 60 * 1000);
        
        const clampedStart = clampToWorkday(start, day, workStartH, workEndH);
        const clampedEnd = clampToWorkday(end, day, workStartH, workEndH);
        
        if (clampedEnd <= clampedStart) continue;
        if (latestEnd && clampedEnd > latestEnd) continue;
        
        const needed = contiguousSlots(clampedStart, minutes);
        const allAvailable = needed.every(([s, e]) => 
          !used.has(`${s.getTime()}-${e.getTime()}`) &&
          availableSlots.some(([as, ae]) => as.getTime() === s.getTime() && ae.getTime() === e.getTime())
        );
        
        if (!allAvailable) continue;
        
        // Mark slots as used
        for (const [s, e] of needed) {
          used.add(`${s.getTime()}-${e.getTime()}`);
        }
        
        plan.blocks.push({
          task_title: task.title,
          start: clampedStart,
          end: clampedEnd
        });
        
        remaining -= minutes;
        placedAny = true;
        break;
      }
      
      if (!placedAny) break;
    }
  }
  
  // Sort blocks by start time and merge adjacent ones
  plan.blocks.sort((a, b) => a.start.getTime() - b.start.getTime());
  return mergeAdjacent(plan);
}

export function energyAlignment(plan: DayPlan, curve?: SlotPoint[]): number {
  if (plan.blocks.length === 0) return 0.0;
  
  const energyLookup = new Map((curve || []).map(([t, e]) => [t.getTime(), e]));
  const hiThresh = 0.75;
  let hi = 0;
  let total = 0;
  
  for (const block of plan.blocks) {
    let t = new Date(block.start);
    t.setSeconds(0, 0);
    
    while (t < block.end) {
      total += 15;
      const energy = energyLookup.get(t.getTime()) || 0.0;
      if (energy >= hiThresh) {
        hi += 15;
      }
      t = new Date(t.getTime() + 15 * 60 * 1000);
    }
  }
  
  return total === 0 ? 0.0 : Math.round((hi / total) * 100) / 100;
}